[ISel] Commute FMUL and inserting zero into vector lane

juliannagele opened the pull request:
When inserting zero into a vector that is the result of a multiplication we can do the insertion earlier, i.e. into an operand of the fmul instead.

juliannagele committed:
Add test for inserting zero into vector lane of fmul result

juliannagele committed:
[ISel] Commute FMUL and inserting zero into vector lane

davemgreen reviewed (commented):
Can you give context why this is better? Thanks

juliannagele commented:
We've observed that doing the insert on an operand of the mul we potentially reduce the critical path, in particular in combination with https://github.com/llvm/llvm-project/pull/146538

fhahn reviewed (commented):
To add maybe a bit more context, commuting the ops may allow the insert to happen earlier, while if we do it after the mul we need to wait for the mul (and its other operand) to be completed.

It should not increase the critial path I think, so in the worst case should not make things worse

davemgreen commented:
OK sounds good. Like you said I don't think this would be worse (and I guess there is a chance to generate fma?). Could we add a test for a case that shows improvement?

How do we know which operand to zero? If it needs trace info then that is only available later in MachineCombiner. (Sometimes I can see that it replies on other DAG optimizations, in which case it sounds OK, it just might not always be reliable).

juliannagele committed:
fixup! [ISel] Commute FMUL and inserting zero into vector lane

juliannagele commented:
Thanks! I've added more tests, including one that is better when doing the the insert early (distilled down from where we found the pattern). I've also added some logic to possibly insert into the other operand and deal with both operands being the same. We could probably add better heuristics for which operand to choose, but if you don't mind I'd leave that as a follow up for now (since it should never be worse as is).

davemgreen reviewed (commented):
I was just checking through - does this need nnan and ninf? And maybe nsz?

juliannagele committed:
Merge branch 'main' into commute-insertzero-fmul

juliannagele committed:
fixup! fixup! [ISel] Commute FMUL and inserting zero into vector lane
